keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -871,32 +880,314 @@ not added to the pod for a long enough time (for example 2 minutes).
+  status:
+    conditions:
+    - status: "True"
+      type: Initialized
+    - status: "False"
+      type: Ready
+    - status: "False"
+      type: ContainersReady
+    - status: "True"
+      type: PodScheduled
+    containerStatuses:
+    - image: non_existing_image_name:102
+      state:
+        waiting:
+          message: Back-off pulling image "non_existing_image_name:102"
+          reason: ImagePullBackOff
+    phase: Pending
+```
+
+The pod is stuck in the Pending and Terminating state. The finalizer
+is not deleted by Job controller as it requires the pod to be in terminal
+phase when `podFailurePolicy` is defined.
+
+**Invalid image reference, pod deleted by scheduler**
+
+1. Create a pod similar to the one above, but extend it with requests to
+ensure it will be preempted by another pod with a higher priority
+2. Create another pod with the critical priority (e.g. `system-node-critical`)
+on the same node so that Kube-scheduler preempts the first pod.
+
+The status looks like in the previous example, but contains the
+`DisruptionTarget=True` condition.
+
+As above, the pod is stuck in the Pending and Terminating state.
+
+**Invalid configMap reference, pod deleted by a user**
+
+1. create a pod with invalid configMap reference, example `yaml`:
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: invalid-configmap-ref
+spec:
+  template:
+    spec:
+      restartPolicy: Never
+      volumes:
+      - name: volume-name
+        configMap:
+          name: invalid-config-name-name
+      containers:
+      - name: invalid-configmap-ref
+        image: centos:7
+        command: ["bash"]
+        args: ["-c", 'echo "Hello world"']
+        volumeMounts:
+        - mountPath: /script_path
+          name: volume-name
+  podFailurePolicy:
+    rules: []
+  backoffLimit: 0
+```
+2. delete the pod with `k delete pods -l job-name=invalid-configmap-ref`
+
+The relevant fields of the pod:
+
+```
+  metadata:
+    deletionTimestamp: "2023-02-03T13:48:14Z"
+    finalizers:
+    - batch.kubernetes.io/job-tracking
+  status:
+    conditions:
+    - status: "True"
+      type: Initialized
+    - status: "False"
+      type: Ready
+    - status: "False"
+      type: ContainersReady
+    - status: "True"
+      type: PodScheduled
+    containerStatuses:
+    - image: centos:7
+      state:
+        terminated:
+          exitCode: 137
+          message: The container could not be located when the pod was terminated
+          reason: ContainerStatusUnknown
+    phase: Pending
+```
+
+As above, the pod is stuck in the Pending and Terminating state.
+
+**Correct config, but image takes long to download, pod deleted by a user in the meanwhile**
+
+This scenario is a little bit different, the config is correct, but the
+pod is deleted by a user while in the `Pending` phase. In that case, the
+pods transition into the `Running` phase and fail soon after. With the proposed
+change the transition will happen earlier, thus saving resources.
+
```

Comment:
nit: I'm not super clear why this scenario will result in phase=Failed being applied, but not the previous scenario. Not needed for the KEP doc itself, but I think we should understand what is going on here...

Reply:
In this scenario the pod actually transitions to the `Running` phase, but the container is killed, respending the grace period. If the container completes with exitCode `0` within the graceful period it actually ends the entire pod in Succeed state, otherwise it ends with exitCode `137`.

Reply:
gotcha, thanks for clarifying this in the doc.